File: packages/cli/src/commands/update.ts
import updateNotifier from 'update-notifier';
import { access } from 'node:fs/promises';
import { constants as fsConstants } from 'node:fs';
import { homedir } from 'node:os';
import { join } from 'node:path';
import { readFile } from 'node:fs/promises';
import { compareVersions, fetchRemoteVersions, getLocalImageVersion } from '../utils/version.js';
import { getConfig } from '../utils/config.js';
import { ensureImage } from '../utils/docker.js';
import { log } from '../utils/ui.js';
import { confirm } from '../utils/prompts.js';

type UpdateOptions = {
  yes?: boolean;
};

function parseCliVersion(packageJsonRaw: string): string {
  const parsed = JSON.parse(packageJsonRaw) as { version: string };
  return parsed.version;
}

export async function notifyCliUpdates(packageJsonPath: string): Promise<void> {
  try {
    const configHome = process.env.XDG_CONFIG_HOME ?? join(homedir(), '.config');
    await access(configHome, fsConstants.W_OK);

    const packageJsonRaw = await readFile(packageJsonPath, 'utf-8');
    const packageVersion = parseCliVersion(packageJsonRaw);
    updateNotifier({ pkg: { name: 'dillinger-gaming', version: packageVersion }, updateCheckInterval: 1000 * 60 * 60 * 6 }).notify();
  } catch {
    // non-blocking
  }
}

export async function updateCheckCommand(): Promise<void> {
  const config = getConfig();
  const remote = await fetchRemoteVersions();
  const localImage = await getLocalImageVersion(config.imageName);

  if (!remote) {
    log.error('Unable to fetch remote versioning.env');
    process.exit(1);
  }

  log.info(`Remote core version: ${remote.coreVersion}`);
  log.info(`Remote script version: ${remote.scriptVersion}`);
  log.info(`Local image version: ${localImage ?? 'not found'}`);

  if (!localImage || compareVersions(localImage, remote.coreVersion) < 0) {
    log.warn('Docker image update available.');
  } else {
    log.success('Docker image is up to date.');
  }
}

export async function updateApplyCommand(options: UpdateOptions): Promise<void> {
  const config = getConfig();
  const remote = await fetchRemoteVersions();

  if (!remote) {
    log.error('Unable to fetch remote versioning.env');
    process.exit(1);
  }

  const imageRef = `${config.imageName}:${remote.coreVersion}`;
  const proceed = options.yes ? true : await confirm(`Pull ${imageRef}?`, true);
  if (!proceed) {
    log.warn('Update cancelled.');
    return;
  }

  await ensureImage(imageRef);
  log.success(`Pulled ${imageRef}`);
}

File: packages/cli_native/src/commands/update.rs
use anyhow::Result;

use crate::utils::{
    config::get_config,
    docker::ensure_image,
    prompts::confirm,
    ui::LOG,
    version::{compare_versions, fetch_remote_versions, get_local_image_version},
};

pub async fn update_check_command() -> Result<()> {
    let config = get_config();
    let remote = fetch_remote_versions().await;
    let local_image = get_local_image_version(&config.image_name)
        .await
        .ok()
        .flatten();

    let remote = match remote {
        Some(r) => r,
        None => {
            LOG.error("Unable to fetch remote versioning.env");
            std::process::exit(1);
        }
    };

    LOG.info(&format!("Remote core version: {}", remote.core_version));
    LOG.info(&format!("Remote script version: {}", remote.script_version));
    LOG.info(&format!(
        "Local image version: {}",
        local_image.as_deref().unwrap_or("not found")
    ));

    if local_image
        .as_ref()
        .map(|l| compare_versions(l, &remote.core_version) < 0)
        .unwrap_or(true)
    {
        LOG.warn("Docker image update available.");
    } else {
        LOG.success("Docker image is up to date.");
    }

    Ok(())
}

pub async fn update_apply_command(yes: bool) -> Result<()> {
    if yes {
        crate::utils::prompts::set_auto_yes(true);
    }

    let config = get_config();
    let remote = fetch_remote_versions().await;

    let remote = match remote {
        Some(r) => r,
        None => {
            LOG.error("Unable to fetch remote versioning.env");
            std::process::exit(1);
        }
    };

    let image_ref = format!("{}:{}", config.image_name, remote.core_version);
    let proceed = confirm(&format!("Pull {}?", image_ref), true)?;
    if !proceed {
        LOG.warn("Update cancelled.");
        return Ok(());
    }

    ensure_image(&image_ref).await?;
    LOG.success(&format!("Pulled {}", image_ref));
    Ok(())
}
